File: frontend/src/features/delta-sync/model/checkpointBuildAbort.ts
/**
 * How the last scheduled checkpoint build aborted before writing anything (issue #224).
 *
 * The backend enum is `CheckpointBuildAbort`; the values are a wire contract. The DTO field is
 * kept a plain string in the Zod schema on purpose — a strict enum would reject the whole
 * sync-state payload over one value added on the server, blanking the Delta Sync tab — so this
 * mapping has to answer for a value it has never seen as well.
 */

import type { SyncSeverity } from './severity';

export interface CheckpointBuildAbortDescription {
  /** Chip / note text. */
  label: string;
  /** Which severity token the reason is painted with. */
  severity: SyncSeverity;
}

const KNOWN: Record<string, CheckpointBuildAbortDescription> = {
  FAILED: { label: 'Build failed', severity: 'critical' },
  FOLD_TOO_LARGE: { label: 'Fold too large', severity: 'critical' },
  FRAME_TOO_LARGE: { label: 'Frame too large', severity: 'critical' },
  // Contention, not a fact about the site — the next tick tries again (#150).
  SCRATCH_FULL: { label: 'Scratch full', severity: 'elevated' },
  FRAME_UNAVAILABLE: { label: 'Could not read frame', severity: 'critical' },
  // Repairs itself once the neighbouring build finishes (#178).
  DEFERRED: { label: 'Build deferred', severity: 'elevated' },
};

/**
 * Describe a recorded scheduled-build abort, or null when the site has none.
 *
 * @param abort the raw value from the sync-state / health projection
 */
export function describeCheckpointBuildAbort(
  abort: string | null | undefined,
): CheckpointBuildAbortDescription | null {
  if (!abort) return null;
  return KNOWN[abort] ?? { label: `Build: ${abort}`, severity: 'elevated' };
}

File: src/main/java/com/bitbi/dfm/delta/domain/CheckpointBuildAbort.java
package com.bitbi.dfm.delta.domain;

/**
 * How a scheduled checkpoint build aborted before writing anything (issue #224).
 *
 * <p>A forced rebuild already leaves {@link CheckpointRebuildOutcome}. The nightly tick did not:
 * every whole-site abort ({@code frame_too_large}, a fold over the heap budget, a deferral behind
 * the process fold budget, an S3 read denial) writes no {@code checkpoints} row and leaves
 * {@code last_checkpoint_seq} at zero, so a first build that has failed thirty nights is
 * indistinguishable from one that is not due yet. This is the abort that tells them apart, and
 * it is written only on the abort path — a healthy build advances the pointer and does not
 * touch these columns.</p>
 *
 * <p>The name is persisted ({@code EnumType.STRING}) and is published on the sync-state REST
 * projection, so these constants are a wire contract: rename one and the frontend stops
 * recognising it.</p>
 *
 * @author Data Forge Team
 * @version 1.0.0
 */
public enum CheckpointBuildAbort {

    /**
     * The build threw for a reason that is not one of the named aborts below. The message is the
     * failure's own text.
     */
    FAILED,

    /**
     * The site's folded state grew past {@code delta.checkpoint.max-fold-bytes} (issue #152). The
     * next tick fails identically, because the site's history has not shrunk.
     */
    FOLD_TOO_LARGE,

    /**
     * The reload frame crossed {@code delta.checkpoint.max-frame-temp-bytes} (issue #153). The
     * next tick fails identically, because the fold has not changed.
     */
    FRAME_TOO_LARGE,

    /**
     * The shared Parquet scratch directory was full (issue #150). Contention, not a fact about
     * the site — the next tick tries again once neighbouring writers finish.
     */
    SCRATCH_FULL,

    /**
     * S3 would not say whether the site's seed frame exists, so the build did nothing
     * (issue #157). Nothing durable changed; the remedy is a bucket policy or IAM grant.
     */
    FRAME_UNAVAILABLE,

    /**
     * Another checkpoint build held the process's fold budget for longer than
     * {@code delta.checkpoint.fold-wait-seconds}, so this one never folded anything
     * (issue #178). It repairs itself on the next tick.
     */
    DEFERRED
}
